### DISCUSSION TITLE
edgedb instance start --dev

### DISCUSSION THREAD

**Discussion by scottmas** [geldata/gel#3974] [Pre-RFC]
As a NoSQL guy who has recently realized SQL is a better fit for most of the applications I'm interested in writing, there's still one thing I miss from my NoSQL database days: the developer experience for adding _non destructive changes_ to my schema. 

For instance, on MongoDB or Firestore if I want to apply some non-destructive change to my db schema (e.g. adding some new optional property or even adding an entirely new table), I literally just make some changes to my application code and I'm off to the races, without any explicit migration step.

I know SQL guys will roll their eyes at my whining at the 30 seconds it will take to run the migration, but it's something I've grown accustomed to. And it's just so stinking convenient and low friction as a developer. 

So my proposal is this (and maybe it's a bad idea - I'm just throwing it out here):

1. Add an optional dev flag to instance start: `edgedb instance start --dev`
2. In `dev` mode, catch all execution errors arising from queries against non existent tables or properties
3. On catch, inspects SDL files for any not-yet-migrated _non destructive changes_ and then force apply them. 
4. If SDL changes were applied, retry the failing query.
5. Et voila, the query succeeds without any explicit migration step by the developer.

With these changes, I could get that workflow back. But what are your thoughts? Is this feasible technically? Is this likely to be a foot gun? I do think a feature like this would be beneficial for people coming from a NoSQL background.

**Comment by tailhook** [comment]
I think we want something like that. But probably just use `inotify` to get schema changes and apply them as they appear in the schema files. Or do you think too much changes will be applied prematurely in this case?

**Reply by haikyuu** [reply]
I guess making the migrations 10x faster with no prompts using a `--dev` mode will solve the issue. 

Or maybe a reverse workflow for prompts where edgedb doesn't prompt in dev mode. But you need to `commit` the dev migrations and it will then ask you the prompts.

Anyway, the problem is real and I'd love to see it addressed somehow 👍

**Reply by scottmas** [reply]
Applying schema changes immediately on write to the SDL files probably would work IF we take away the non-destructive requirement, since there's likely to be quite a bit of flux in the first little bit while ironing out the schema changes. It's quite likely that the dev will make some SDL changes that are then reverted shortly thereafter, necessitating a destructive change and ruining the seamless workflow we're after. 

The best of both worlds would be have a log of changes and only allow destructive changes if the destructive change would be applied to a table or column created in the past 2 hours or something like that. But that's perhaps getting too clever. I'm just thinking of ways we might mitigate inadvertent loss of dev data. But it's probably not too big of a deal.

I'd also be okay with a `edgedb migration apply --dev` flag or the like. Slightly higher friction but not too bad, and more explicit.

**Reply by tailhook** [reply]
I think the plan is to start with applying destructive changes too. The only "magic" thing is that we will not generate a separate migration file for every single change. So when you want to commit `dbschema/migrations`, there is only a single file.

I think in most dev workflows loosing some values is not a big deal. We may do some backups later to make this even nicer.

I'll try to outline more concrete proposal soon.

**Reply by scottmas** [reply]
What approach are we currently thinking on implementation? Lazy schema changes on error? On SDL changes  using `inotify`? A manual cli command?

**Reply by tailhook** [reply]
SDL changes using `inotify` is somethink we consider first, I think.
